nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/field/ObjectLocalDateTimeFieldConverter.java

```diff
@@ -51,10 +51,13 @@ public LocalDateTime convertField(final Object field, final Optional<String> pat
             final Instant instant = Instant.ofEpochMilli(date.getTime());
             return ofInstant(instant);
         }
-        if (field instanceof Number) {
-            final Number number = (Number) field;
-            final Instant instant = Instant.ofEpochMilli(number.longValue());
-            return ofInstant(instant);
+        if (field instanceof final Number number) {
+            // If value is a floating point number, we consider it as seconds since epoch plus a decimal part for fractions of a second.
+            if (field instanceof Double || field instanceof Float) {
+                // Use long value as seconds; use everything after it as fractions of a second and multiply by 1_000_000_000 to get nanoseconds.
+                return toLocalDateTime(number.longValue(), (long) ((number.doubleValue() - number.longValue()) * 1_000_000_000));
```

Comment:
Recommend breaking out nanosecond assignment to a separate line for readability, as it is a bit difficult to follow the levels of nested calculation.